File: 1763-longest-nice-substring/1763-longest-nice-substring.py
class Solution:
    def isNice(self,substring) -> bool:
        lower = [False] * 26
        upper = [False] * 26

        for char in substring:
            if char.islower():
                lower[ord(char) - ord('a')] = True
            elif char.isupper():
                upper[ord(char) - ord('A')] = True

        for i in range(26):
            if lower[i] != upper[i]:
                return False
        return True
    
    def longestNiceSubstring(self, s: str) -> str:
        if not s or len(s) < 2:
            return ""

        longestSubstring = ""
        n = len(s)
        for i in range(n):
            for j in range(i + 1, n + 1):
                currentSubstring = s[i:j]
                if self.isNice(currentSubstring) and len(currentSubstring) > len(longestSubstring):
                    longestSubstring = currentSubstring

        return longestSubstring

File: 1763-longest-nice-substring/1763-longest-nice-substring.java
class Solution {
    private boolean isNice(String s){
        boolean[] lower=new boolean[26];
        boolean[] upper=new boolean[26];
        for(char c : s.toCharArray()){
            if(Character.isLowerCase(c)){
                lower[c-'a']=true;
            }
            else if (Character.isUpperCase(c)) {
                upper[c - 'A'] = true;
            }
        }
        for (int i = 0; i < 26; i++) {
            if (lower[i] != upper[i]) {
                return false;
            }
        }
        return true;        
    }
    public String longestNiceSubstring(String s) {
        if(s.isEmpty() || s.length()<2) return "";
        String longestSubstring="";
        // Generate all possible substrings
        for (int i = 0; i < s.length(); i++) {
            for (int j = i + 1; j <= s.length(); j++) {
                String substring = s.substring(i, j);
                if (isNice(substring) && substring.length() > longestSubstring.length()) {
                    longestSubstring = substring;
                }
            }
        }
        
        return longestSubstring;        
    }
}